File: app/controllers/parser_controller.rb
class ParserController < ApplicationController
  include ApplicationHelper
  include ParserHelper
  # Specify :command as a model here; otherwise in ParserHelper, Command#find
  # raises an exception: "uninitialized constant". See "Application helper not recognising model",
  # http://wrath.rubyonrails.org/pipermail/rails/2005-February/003326.html
  # [Jon Aquino 2005-07-16]
  model :command
  attr_reader :last_url

  def parse
    if empty_string_if_nil(@params[:command]).strip == '' then
      redirect_to(:action => 'index')
      return
    end
    if @params[:show_user_agent] then
      render_text request.user_agent
      return
    end
    # According to the access log, Yahoo Pipes seems to be bringing the site down. [Jon Aquino 2009-07-03]
    if request.user_agent =~ /Yahoo Pipes/ then
      render :text => "YubNub is currently blocking Yahoo Pipes. Contact jonathan.aquino@gmail.com for more info.", :status => 403
      return
    end
    # Cache the url, for testing [Jon Aquino 2005-06-19]
    @last_url = parse_with_substitutions(@params[:command], @params[:default])
    redirect_to_url @last_url
  end
  def url
    render_text parse_with_substitutions(@params[:command], nil)
  end
  def index
    @page_title = "YubNub"
  end
  def uptime
    Command.find_first()
    render_text "success"
  end
end

File: app/controllers/ParserController.php
<?php

/**
 * Parses commands entered by the user.
 */
class ParserController extends Controller {

    /**
     * Parses a command entered by the user.
     *
     * Expected GET variables:
     *     - command - the command string, e.g., g porsche
     */
    public function action_parse() {
        $command = isset($_GET['command']) ? trim($_GET['command']) : null;
        $defaultCommand = isset($_GET['default']) ? trim($_GET['default']) : 'g';
        if (strlen($command) == 0) {
            $this->redirectTo('/');
        }
        // According to the access log, Yahoo Pipes seems to be bringing the site down. [Jon Aquino 2009-07-03]
        if (strpos($_SERVER['HTTP_USER_AGENT'], 'Yahoo Pipes') > -1) {
            header('HTTP/1.0 403 Forbidden');
            echo 'YubNub is currently blocking Yahoo Pipes. Contact jonathan.aquino@gmail.com for more info.';
            exit;
        }
        $parser = new Parser(new CommandStore($this->pdoSingleton->getPdo()));
        $url = $parser->parse($command, $defaultCommand);
        if (strpos($url, '[post]') !== false) {
            $url = str_replace('[post]', '', $url);
            $this->render('get2post', array(
                'url' => new Url($url)
            ));
        }
        $this->redirectTo($url);
    }

    /**
     * Outputs the URL that the given command resolves to.
     */
    public function action_url() {
        $command = isset($_GET['command']) ? trim($_GET['command']) : null;
        $parser = new Parser(new CommandStore($this->pdoSingleton->getPdo()));
        header('Content-Type: text/plain');
        echo $parser->parse($command, null);
    }

    /**
     * Outputs 'success' if Yubnub seems to be healthy.
     */
    public function action_uptime() {
        $commandStore = new CommandStore($this->pdoSingleton->getPdo());
        $commands = $commandStore->findCommands(array('start' => 0, 'count' => 1,
                'orderBy' => 'creation_date DESC'));
        if (count($commands) == 1) {
            header('Content-Type: text/plain');
            echo 'success';
        }
    }

}
